### ISSUE TITLE
Basic XSS protection/HTML sanitization for all parsed content

### ISSUE THREAD

**Issue by julianlam** [NodeBB/NodeBB#7872]
Markdown plugin provides html sanitization and XSS protection out of the box. If you don't use Markdown, you probably use composer-quill or composer-redactor, and those handle sanitization on their own.

However, in the rare event that all of these plugins get disabled, then you have no protection whatsoever. It's rare enough that it has not been an issue, but some basic protection would not go amiss.

Starting in v1.13.x, we will bundle HTML sanitization into core so that all `filter:parse.*` hooks will be sanitized as the last thing. It shouldn't affect most plugins unless you use rather esoteric html tags, but should help guard against malicious behaviour.

If you happen to use a plugin that gets adversely affected by this plugin, you'll want to hook into the new hook `filter:sanitize.config`, in order to adjust the configuration.

**Comment by NormandoHall** [comment]
Please, this break the youtube lite plugin. How can I add extra tags using the new hook in my code?

These are the plugin hooks used:

```
{ "hook": "filter:parse.raw", "method": "parseRaw", "priority": 10 },
{ "hook": "filter:parse.post", "method": "parsePost", "priority": 10 },
{ "hook": "filter:admin.header.build", "method": "addAdminNavigation" },
{ "hook": "static:app.load", "method": "init" }	
```

I suppose need to add something like this:

`{ "hook": "filter:sanitize.config", "method": "sanitizeConfig", "priority": 10 }`

in the plugin method sanitizeConfig, but I am new to plugin develop.

These are the tags plugin use:

```
content += '<div class="js-lazyYT lazyYT-container" data-youtube-id="' + videoId + '" data-width="640" data-height="360" data-parameters="' + queryString + '" style="width: 640px; padding-bottom: 360px;">' +
                            '\n <div class="ytp-thumbnail lazyYT-image-loaded" style="background-image: url(&quot;' + img.url + '&quot;);">' +
                            '\n  <button class="ytp-large-play-button ytp-button" tabindex="23" aria-live="assertive" style="transform: scale(0.85);" onclick="$(this).lazyYT(this);return false;">' +
                            '\n   <svg height="100%" version="1.1" viewBox="0 0 68 48" width="100%">' +
                            '\n    <path class="ytp-large-play-button-bg" d="m .66,37.62 c 0,0 .66,4.70 2.70,6.77 2.58,2.71 5.98,2.63 7.49,2.91 5.43,.52 23.10,.68 23.12,.68 .00,-1.3e-5 14.29,-0.02 23.81,-0.71 1.32,-0.15 4.22,-0.17 6.81,-2.89 2.03,-2.07 2.70,-6.77 2.70,-6.77 0,0 .67,-5.52 .67,-11.04 l 0,-5.17 c 0,-5.52 -0.67,-11.04 -0.67,-11.04 0,0 -0.66,-4.70 -2.70,-6.77 C 62.03,.86 59.13,.84 57.80,.69 48.28,0 34.00,0 34.00,0 33.97,0 19.69,0 10.18,.69 8.85,.84 5.95,.86 3.36,3.58 1.32,5.65 .66,10.35 .66,10.35 c 0,0 -0.55,4.50 -0.66,9.45 l 0,8.36 c .10,4.94 .66,9.45 .66,9.45 z" fill="#1f1f1e" fill-opacity="0.9">' +
                            '\n    </path>' +
                            '\n    <path d="m 26.96,13.67 18.37,9.62 -18.37,9.55 -0.00,-19.17 z" fill="#fff">' +
                            '\n    </path>' + 
                            '\n    <path d="M 45.02,23.46 45.32,23.28 26.96,13.67 43.32,24.34 45.02,23.46 z" fill="#ccc">' +
                            '\n    </path>' +
                            '\n   </svg>' +
                            '\n  </button>' +
                            '\n </div>' +
                            '\n</div>';
```

**Comment by barisusakli** [comment]
You need to add the required tags and attributes in that html to your plugin. It would look like this
```
myPlugin.filterSanitizeConfig = function (sanitizeConfig, callback) {
    sanitizeConfig.allowedTags.push('svg', 'path');
    sanitizeConfig.allowedAttributes.svg = sanitizeConfig.allowedAttributes.svg || [];
    sanitizeConfig.allowedAttributes.svg.push('height', 'weight' /* more attributes if necessary */);
    sanitizeConfig.allowedAttributes.path = sanitizeConfig.allowedAttributes.path || [];
    sanitizeConfig.allowedAttributes.path.push('d', 'fill', 'fill-opacity' /* more attributes if necessary */);
};
```

Adjust as necessary until it renders properly.

**Comment by NormandoHall** [comment]
Thank you so much @barisusakli 

I added at the end of the function: `callback(null, sanitizeConfig);`

One thing more. There are some attributes that are camelcase, like "viewBox". When I whitelisted "viewBox" this not appear. If I change plugin code to lowercase "viewbox" and added to sanitizer funcion the attribute also in lowercase "viewbox" it appear in cameCase! Strange "viewBox". It is no a problem. I tell you only for your information.

This (how to whitelist html tags) can be added to the Developers FAQ.

Thanks!
